tests/std/tests/P0896R4_counted_iterator/test.cpp

```diff
@@ -0,0 +1,185 @@
+                iter++;
+                assert(iter.count() == 1);
+                assert(*iter == 2);
+            }
+            if constexpr (bidirectional_iterator<Iter>) {
+                {
+                    // pre decrement
+                    auto iter = counted_iterator<Iter>{Iter{input + 2}, 3};
+                    --iter;
+                    assert(iter.count() == 4);
+                    assert(*iter == 2);
+                }
+                { // post decrement
+                    auto iter = counted_iterator<Iter>{Iter{begin(input) + 2}, 3};
+                    iter--;
+                    assert(iter.count() == 4);
+                    assert(*iter == 2);
+                }
+            }
+            if constexpr (random_access_iterator<Iter>) {
+                { // increment by n
+                    auto iter = counted_iterator<Iter>{Iter{input}, 5};
+                    auto ref  = iter + 3;
+                    assert(iter.count() == 5);
+                    assert(*iter == 1);
+                    assert(iter.count() == ref.count() + 3);
+                    assert(iter.base().peek() == prev(ref.base().peek(), 3));
+                }
+                { // post increment by n
+                    auto iter = counted_iterator<Iter>{Iter{input}, 5};
+                    auto ref  = 3 + iter;
+                    assert(iter.count() == 5);
+                    assert(*iter == 1);
+                    assert(iter.count() == ref.count() + 3);
+                    assert(iter.base().peek() == prev(ref.base().peek(), 3));
+                }
+                { // increment assign
+                    auto iter = counted_iterator<Iter>{Iter{input}, 5};
+                    auto ref  = iter;
+                    iter += 3;
+                    assert(iter.count() == 2);
+                    assert(*iter == 4);
+                    assert(iter.count() == ref.count() - 3);
+                    assert(iter.base().peek() == next(ref.base().peek(), 3));
+                }
+                { // decrement by n
+                    auto iter = counted_iterator<Iter>{Iter{begin(input) + 2}, 3};
+                    auto ref  = iter - 2;
+                    assert(iter.count() == 3);
+                    assert(*iter == 3);
+                    assert(iter.count() == ref.count() - 2);
+                    assert(iter.base().peek() == next(ref.base().peek(), 2));
+                }
+            }
+            { // difference
+                auto iter1 = counted_iterator<Iter>{Iter{input + 1}, 2};
+                auto iter2 = counted_iterator<Iter>{Iter{input}, 3};
+
+                const same_as<iter_difference_t<Iter>> auto diff1 = iter1 - iter2;
+                assert(diff1 == 1);
+
+                const same_as<iter_difference_t<Iter>> auto diff2 = iter2 - iter1;
+                assert(diff2 == -1);
+            }
+            // TODO different type that satisfies common_with

```

Comment:
Ditto "Should we make `test::iterator<stuff>` convert to `test::iterator<const stuff>`?

Reply:
Done